File: 04/main.py
import sys
import pprint


def check_win(board):
    for x in range(len(board)):
        if all([n == -1 for n in board[x]]):
            return True

    for y in range(len(board)):
        if all([board[x][y] == -1 for x in range(len(board))]):
            return True

    return False


def sum_board(board):
    ret = 0
    for x in range(len(board)):
        for y in range(len(board)):
            if board[x][y] != -1:
                ret += board[x][y]
    return ret


def replace_in_board(board, n):
    for x in range(5):
        for y in range(5):
            if board[x][y] == n:
                board[x][y] = -1
    return board


def parse_boards():
    with open("input.txt") as f:
        numbers = [int(x) for x in f.readline().strip().split(",")]
        boards = f.read().replace("  ", " ").split("\n\n")

    boards = [[[int(x) for x in line.split(" ") if x] for line in board.split("\n") if line] for board in boards]
    return numbers, boards


numbers, boards = parse_boards()
for n in numbers:
    for i in range(len(boards)):
        boards[i] = replace_in_board(boards[i], n)

        if check_win(boards[i]):
            pprint.pprint(boards[i])
            print(sum_board(boards[i]) * n)
            sys.exit(0)

File: 04/main.cpp
#include "../common.h"
#include <iostream>

#define Board vector<vector<int>>

#define BOARD_SIZE 5

using namespace std;

Board replace_in_board(Board board, int n)
{
    for (int x = 0; x < BOARD_SIZE; x++) {
        for (int y = 0; y < BOARD_SIZE; y++) {
            if (board[x][y] == n) {
                board[x][y] = 0;
            }
        }
    }
    return board;
}

bool check_win(Board board)
{
    for (int x = 0; x < BOARD_SIZE; x++) {
        int sum = 0;
        for (int n : board[x])
            sum += n;
        if (sum == 0)
            return true;
    }

    for (int y = 0; y < BOARD_SIZE; y++) {
        int sum = 0;
        for(int x=0;x<BOARD_SIZE;x++)
            sum += board[x][y];
        if (sum == 0)
            return true;
    }

    return false;
}

int sum_board(Board board)
{
    int sum = 0;
    for(vector<int> line : board) {
        for(int e : line) {
            sum += e;
        }
    }
    return sum;
}

int main()
{
    ifstream fp("input.txt");

    string nums_raw;
    getline(fp, nums_raw);
    vector<string> nums = split(nums_raw, ",");

    int num;
    vector<Board> boards;
    while (!fp.eof()) {
        Board board;

        for (int i = 0; i < BOARD_SIZE; i++) {
            vector<int> line;
            for (int j = 0; j < BOARD_SIZE; j++) {
                fp >> num;
                line.push_back(num);
            }
            board.push_back(line);
        }
        boards.push_back(board);
    }
    boards.pop_back();

    for (string sn : nums) {
        int n = stoi(sn);
        for (int i = 0; i < boards.size(); i++) {
            boards[i] = replace_in_board(boards[i], n);

            if (check_win(boards[i])) {
                cout << sum_board(boards[i]) * n << "\n";
                return 0;
            }
        }
    }
}